h/static/scripts/base/controller.js

```diff
@@ -86,6 +106,95 @@ class Controller {
   forceUpdate() {
     this.update(this.state, this.state);
   }
+
+  /**
+   * Replace the HTML content of the element with an updated version from the
+   * server.
+   *
+   * Since this replaces the entire element, the current controller will not
+   * receive events for the replaced element. Instead the new controller for the
+   * replaced element is returned and can be used to transfer state from the
+   * existing controller.
+   *
+   * @param {string} html - The new markup for the element
+   * @return {Controller} - The new controller instance of the same type as this
+   *         controller on the element that replaces `this.element`
+   */
+  reload(html) {
+    if (!this.options.reload) {
+      throw new Error('No reload() function supplied to controller constructor');
+    }
+
+    var newElement = this.options.reload(this.element, html);
+    var ctrl = (newElement.controllers || []).find(ctrl => {
+      return ctrl instanceof this.constructor;
+    });
+
+    this.trigger('reload', {newController: ctrl});
+
+    return ctrl;
+  }
+
+  /**
+   * Return controllers attached to child elements which are instances of
+   * `ControllerClass`
+   *
+   * @return {Array<ControllerClass>} Array of controllers attached to children
+   *         of `this.element` which are instances of `ControllerClass`
+   */
+  childControllers(ControllerClass) {
+    var children = this.element.querySelectorAll('*');
```

Comment:
This doesn't seem right to me. This code means that if we were to have a single controller with a single child controller, the code would slow down linearly in the size of the DOM of the parent controller.

As suggested earlier, let's either remove the concept of child controllers for now, or explicitly keep track of children.